README.md
tfluna
======

![tfluna icon](media/icon.png)

Forward TF-Luna LIDAR sensor events over OSC, UDP, or Thingsboard.

This code base has been developed by [ZKM | Hertz-Lab](https://zkm.de/en/about-the-zkm/organization/hertz-lab) as part of the project [»The Intelligent Museum«](#the-intelligent-museum). 

Copyright (c) 2022 ZKM | Karlsruhe.  
Copyright (c) 2022 Dan Wilcox.  
Copyright (c) 2020 Daniel Heiss <heiss@zkm.de>  
Copyright (c) 2020 Marc Schütze <mschuetze@zkm.de>  

BSD Simplified License.

Description
-----------

This script sends TF-Luna 1D ToF (Time of Flight) LIDAR sensor proximity distance measurements over OSC (default) or UDP
and optional "isThere" presence events to a [ThingsBoard URL](https://thingsboard.io).

The distance measurement format is an integer in cm (default) or a normalized float (inverted, 1 near to 0 far).

Dependencies
------------

* Python 3
* [pyserial library](https://github.com/pyserial/pyserial/)
* [python-osc library](https://github.com/attwad/python-osc)
* [requests library](https://github.com/psf/requests)

Setup
-----

Install Python 3, if not already available. For instance, on macOS using [Homebrew](http://brew.sh):

```shell
brew install python3
```

Create a virtual environment and install the script's dependencies:

```shell
make
```

### Enable Raspberry Pi serial port for TF-Luna

The TF-Luna can be read over one of the built-in serial ports connected to the GPIO pins. Older boards (RPi 3) can use UART 1 or 2 while newer boards (RPi 4+) have additional UARTs available.

https://www.raspberrypi.com/documentation/computers/configuration.html#configuring-uarts

![pi tfluna wiring diagram](media/pi%20tfluna%20wiring%20diagram.png)

_Note: RX/TX wires may be different colors._

#### UART 1

RPi 3 or RPi 4

GPIO pin connections
* GPIO 14: sensor TX
* GPIO 15: sensor RX
* 5V+: sensor 5V (red)
* GND: sensor GND (black)

This serial port is used for bluetooth by default. Disable this service and ensure the mini UART uses UART 0 by adding the following to `/boot/config.txt`:

~~~
dtoverlay=pi3-disable-bt,pi3-miniuart-bt
~~~

Then reboot:

~~~
sudo reboot
~~~

The serial port should now be available as `/dev/ttyAMA0`.

If the port cannot be read (permissions issues), try the following [suggested on Stack Exchange](https://raspberrypi.stackexchange.com/a/48258 ):

1. Remove anything that looks like `console=serial0,115200
, console=tty1` in `/boot/cmdline.txt` file
2. Disable the default tty and give the user access to the `dialout` group
~~~
sudo systemctl mask serial-getty@ttyAMA0.service
sudo adduser $USER dialout
~~~
3. Reboot
~~~
sudo reboot
~~~

#### UART 3

RPi 4+

GPIO pin connections
* GPIO 4: sensor TX
* GPIO 5: sensor RX
* 5V+: sensor 5V (red)
* GND: sensor GND (black)

This UART is disabled by default. Enable it by adding the following to `/boot/config.txt`

~~~
dtoverlay=uart3
~~~

Then reboot:

~~~
sudo reboot
~~~

The serial port should now be available as `/dev/ttyAMA1`.

### USB to TTL dongle

Some versions of the TF-Luna come with an included "USB to TTL" dongle which is essentially a USB serial port adapter with bare pins for the sensor. 

On macOS, once the adapter is plugged in, it's serial port should be available at some variation of `/dev/tty.usbserial-*` such as `/dev/tty.usbserial-410` or `/dev/tty.usbserial-510`.

Running
-------

Start reading the sensor at serial port `DEV` on the commandline via the virtual environment wrapper script:

    ./tfluna DEV

For example:

    ./tfluna /dev/ttyAMA1

It can simply sit in the background and send events over a networking connection. To configure the send address and ports as well as various other options, see the commandline argument help for tfluna by running:

    ./tfluna -h

Defaults are:

* OSC to 127.0.0.1 on port 5005
* message format: `/tfluna distance`
* distance format: centimeters
* max distance: 200 cm

Messages are sent over UDP either formatted as OSC messages (default) or raw UDP packets by using the `--udp` option. The message name is `tfluna` by default but can be overridden by using the `--message` option:

Multiple instances can be run at the same time by providing a different serial port DEV argument. Additionally, an optional device identifier can be added to the message to differentiate between instances via the `--id` option. 

~~~
Message format: message [id] distance

  Ex. OSC: "/tfluna" distance

  Ex. UDP: "tfluna" distance
~~~

Distance measurements can be sent in a normalized format by using the `--normalize` option which uses an inverted range of 1 near and 0 far to the max distance which can also be specified in cm with the `--max-distance` option.

Optionally, a boolean "isThere" event can be sent to a ThingsBoard URL whenever
someone moves in front or away from the sensor:

~~~
$ --tb-url http://board.mydomain.com/api/v1/TOKEN/telemetry
TB: {"isThere": 0}
~~~

The default "isThere" message name can be overridden via --tb-message:

~~~
$ --tb-url http://board.mydomain.com/api/v1/TOKEN/telemetry \\
--tb-message proximity
TB: {"proximity": 0}
~~~

Test Client
-----------

A simple test client is provided which receives UDP messages on the default 127.0.0.1 address and port 5005 and prints their contents to the console.

To use, start the client:

    ./tests/udpreceive.py

Next, start tfluna in another shell:

    ./tfluna --udp

_Note: This client can also receive and print OSC messages, however the message content is printed in it's raw byte format._

The Intelligent Museum
----------------------

An artistic-curatorial field of experimentation for deep learning and visitor participation

The [ZKM | Center for Art and Media](https://zkm.de/en) and the [Deutsches Museum Nuremberg](https://www.deutsches-museum.de/en/nuernberg/information/) cooperate with the goal of implementing an AI-supported exhibition. Together with researchers and international artists, new AI-based works of art will be realized during the next four years (2020-2023).  They will be embedded in the AI-supported exhibition in both houses. The Project „The Intelligent Museum” is funded by the Digital Culture Programme of the [Kulturstiftung des Bundes](https://www.kulturstiftung-des-bundes.de/en) (German Federal Cultural Foundation) and funded by the [Beauftragte der Bundesregierung für Kultur und Medien](https://www.bundesregierung.de/breg-de/bundesregierung/staatsministerin-fuer-kultur-und-medien) (Federal Government Commissioner for Culture and the Media).

As part of the project, digital curating will be critically examined using various approaches of digital art. Experimenting with new digital aesthetics and forms of expression enables new museum experiences and thus new ways of museum communication and visitor participation. The museum is transformed to a place of experience and critical exchange.

![Logo](media/Logo_ZKM_DMN_KSB.png)


CHANGES.txt
1.0.1: 2022 Jul 26

* fixed wrapper script string escaping 

1.0.0: 2022 Mar 09

initial version split out of museum-label project repo, version 1.1.0


tests/udpreceive.py
#! /usr/bin/env python3
#
# Copyright (c) 2020 ZKM | Museums Technik
# Daniel Heiss <heiss@zkm.de>
# Marc Schütze <mschuetze@zkm.de>
#
# BSD Simplified License.
# For information on usage and redistribution, and for a DISCLAIMER OF ALL
# WARRANTIES, see the file, "LICENSE.txt," in this distribution.
#
# ref: http://sfriederichs.github.io/how-to/python/udp/2017/12/07/UDP-Communication.html

import socket
from time import sleep

UDP_IP = "" # any address
UDP_PORT = 5005
is_running = True

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) # UDP
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

sock.bind((UDP_IP, UDP_PORT))
sock.setblocking(0)

try:
    while is_running:
        try:
            data, addr = sock.recvfrom(1024)
            print("received message: %s" % data)
        except socket.error:
            pass
        sleep(0.1)
except KeyboardInterrupt:
    is_running = False


luna/TFLuna.py
#! /usr/bin/env python3
#
# Copyright (c) 2020 ZKM | Museums Technik
# Daniel Heiss <heiss@zkm.de>
# Marc Schütze <mschuetze@zkm.de>
#
# Copyright (c) 2021 ZKM | Hertz-Labor
# Dan Wilcox <dan.wilcox@zkm.de>
# Paul Bethge <bethge@zkm.de>
#
# BSD Simplified License.
# For information on usage and redistribution, and for a DISCLAIMER OF ALL
# WARRANTIES, see the file, "LICENSE.txt," in this distribution.
#
# This code has been developed at ZKM | Hertz-Lab as part of „The Intelligent
# Museum“ generously funded by the German Federal Cultural Foundation.

import serial
import time

##### math

# map a value within an input range min-max to an output range
def map_value(value, inmin, inmax, outmin, outmax):
    # the following is borrowed from openframeworks ofMath.cpp
    if abs(inmin - inmax) < 0.0001:
        return outmin
    else:
        outval = ((value - inmin) / (inmax - inmin) * (outmax - outmin) + outmin)
        if outmax < outmin:
            if outval < outmax:
                outval = outmax
            elif outval > outmin:
                outval = outmin
        else:
            if outval > outmax:
                outval = outmax
            elif outval < outmin:
                outval = outmin
        return outval

# clamp a value to an output range min-max
def clamp_value(value, outmin, outmax):
    return max(min(value, outmax), outmin)

##### TFLuna

class TFLuna:

    # init with dev path/name and optional baud rate, device identifier, or verbosity
    def __init__(self, dev, rate=115200, devid=None, verbose=True):
        self.serial = serial.Serial(dev, rate)
        self.devid = devid      # optional device identifier, unrelated to serial dev
        self.prev_distance = 0  # previous distance in cm
        self.max_distance = 200 # distance threshold in cm
        self.epsilon = 2        # change threshold in cm
        self.interval = 0.1     # sleep idle time in s
        self.normalize = False  # normalize measured distance?
        self.is_running = True
        self.senders = []
        self.verbose = verbose
        if self.verbose:
            print(f"tfluna: created {dev} {rate}")

    # add a distance sender which implements the following method:
    # send(self, distance, tfluna)
    def add_sender(self, sender):
        self.senders.append(sender)

    # open serial port for reading
    def open(self):
        if not self.serial.is_open:
            self.serial.open()
            if self.verbose:
                print("tfluna: open")

    # close serial port
    def close(self):
        if self.serial.is_open:
            self.serial.close()
            if self.verbose:
                print("tfluna: close")

    # start synchronous run loop
    def start(self):
        self.is_running = True
        if self.verbose:
            print("tfluna: start")
        while self.is_running:
            self.update()
            time.sleep(self.interval)
        self.is_running = False
        if self.verbose:
            print("tfuna: stop")

    # stop synchronous run loop
    def stop(self):
        self.is_running = False

    # read one chunk of data, parse into distance value, and send
    def update(self):
        count = self.serial.in_waiting
        if count > 4:
            recv = self.serial.read(4)
            self.serial.reset_input_buffer()

            # check if input is valid
            if recv[0] == 89 and recv[1] == 89:
                #print(f"{str(recv[0])} {str(recv[1])} {str(recv[2])} {str(recv[3])}")

                # interpret upper two bytes as one 16 bit int
                low = int(recv[2])
                high = int(recv[3])
                distance = clamp_value(low + high * 256, 0, self.max_distance)
            
                # ignore if difference from prev value is too small
                if abs(distance - self.prev_distance) < self.epsilon:
                    return
                self.prev_distance = distance

                # map to normalized range? 0 far to 1 near
                if self.normalize:
                    distance = map_value(distance, 0, self.max_distance, 1, 0)

                # send
                if self.verbose:
                    print(f"tfluna: {distance}")
                for sender in self.senders:
                    sender.send(distance, self)

    # print settings
    def print(self):
        print(f"tfluna: device id {self.devid}")
        print(f"tfluna: max distance {self.max_distance}")
        print(f"tfluna: epsilon {self.epsilon}")
        print(f"tfluna: interval {self.interval}")
        print(f"tfluna: normalize {self.normalize}")


luna/TBSender.py
#! /usr/bin/env python3
#
# Copyright (c) 2021 ZKM | Hertz-Labor
# Dan Wilcox <dan.wilcox@zkm.de>
# Paul Bethge <bethge@zkm.de>
#
# BSD Simplified License.
# For information on usage and redistribution, and for a DISCLAIMER OF ALL
# WARRANTIES, see the file, "LICENSE.txt," in this distribution.
#
# This code has been developed at ZKM | Hertz-Lab as part of „The Intelligent
# Museum“ generously funded by the German Federal Cultural Foundation.

import requests
from concurrent.futures import ThreadPoolExecutor

# ThingsBoard message sender
class TBSender:

    # init with thingsboard url (including access token) and optional threading
    def __init__(self, url, threading=False):
        self.url = url
        self.message = "isThere" # message name
        self.is_there = False # is someone/something there? ie. blocking sensor
        self.verbose = False
        self.pool = None
        if(threading):
            self.pool = ThreadPoolExecutor(16) # worker thread pool

    # send event on change
    def send(self, distance, tfluna):
        is_there = distance < tfluna.max_distance
        if is_there == self.is_there:
            return
        self.is_there = is_there
        if self.verbose:
            print(f"tb sender: {self.message} {self.is_there}")
        payload = {self.message: int(self.is_there == True)}
        if self.pool:
            self.pool.submit(self._post, (payload))
        else:
            self._post(payload)

    # print settings
    def print(self):
        print(f"tb sender: {self.url}")
        print(f"tb sender: sending {self.message}")

    # send the post request with json payload
    def _post(self, payload):
        try:
            req = requests.post(self.url, json=payload)
            if req.status_code != 200:
                print(f"tb sender: send error {req.status_code}")
        except Exception as e:
            print(f"tb sender: send error: {e}")


luna/OSCSender.py
#! /usr/bin/env python3
#
# Copyright (c) 2021 ZKM | Hertz-Labor
# Dan Wilcox <dan.wilcox@zkm.de>
# Paul Bethge <bethge@zkm.de>
#
# BSD Simplified License.
# For information on usage and redistribution, and for a DISCLAIMER OF ALL
# WARRANTIES, see the file, "LICENSE.txt," in this distribution.
#
# This code has been developed at ZKM | Hertz-Lab as part of „The Intelligent
# Museum“ generously funded by the German Federal Cultural Foundation.

from pythonosc import udp_client

# Open Sound Control message sender
class OSCSender:

    # init with address pair: (host, port)
    def __init__(self, addr):
        host,port = addr
        self.client = udp_client.SimpleUDPClient(host, port)
        self.addr = addr
        self.address = "/distance" # OSC address

    # send distance value
    # TODO: ignore sending until timeout after a send error?
    def send(self, distance, tfluna):
        args = [tfluna.devid, distance] if tfluna.devid else distance
        try:
            self.client.send_message(self.address, args)
        except Exception as e:
            print(f"osc sender: send error: {e}")

    # print settings
    def print(self):
        host,port = self.addr
        print(f"osc sender: {host} {port}")
        print(f"osc sender: sending {self.address}")


luna/UDPSender.py
#! /usr/bin/env python3
#
# Copyright (c) 2020 ZKM | Museums Technik
# Daniel Heiss <heiss@zkm.de>
# Marc Schütze <mschuetze@zkm.de>
#
# Copyright (c) 2021 ZKM | Hertz-Labor
# Dan Wilcox <dan.wilcox@zkm.de>
# Paul Bethge <bethge@zkm.de>
#
# BSD Simplified License.
# For information on usage and redistribution, and for a DISCLAIMER OF ALL
# WARRANTIES, see the file, "LICENSE.txt," in this distribution.
#
# This code has been developed at ZKM | Hertz-Lab as part of „The Intelligent
# Museum“ generously funded by the German Federal Cultural Foundation.

import socket

# raw UDP message sender
class UDPSender:

    # init with address pair: (host, port)
    def __init__(self, addr):
        self.client = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) # UDP
        self.client.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.addr = addr
        self.message = "distance" # message text

    # send distance value
    # TODO: ignore sending until timeout after a send error?
    def send(self, distance, tfluna):
        message = self.message + " " + str(tfluna.devid) if tfluna.devid else self.message
        message = (message + " " + str(distance)).encode()
        try:
            self.client.sendto(message, self.addr)
        except Exception as e:
            print(f"udp sender: send error: {e}")

    # print settings
    def print(self):
        host,port = self.addr
        print(f"udp sender: {host} {port}")
        print(f"udp sender: sending {self.message}")


tfluna.py
#! /usr/bin/env python3
#
# Copyright (c) 2020 ZKM | Museums Technik
# Daniel Heiss <heiss@zkm.de>
# Marc Schütze <mschuetze@zkm.de>
#
# Copyright (c) 2021 ZKM | Hertz-Labor
# Dan Wilcox <dan.wilcox@zkm.de>
# Paul Bethge <bethge@zkm.de>
#
# BSD Simplified License.
# For information on usage and redistribution, and for a DISCLAIMER OF ALL
# WARRANTIES, see the file, "LICENSE.txt," in this distribution.
#
# This code has been developed at ZKM | Hertz-Lab as part of „The Intelligent
# Museum“ generously funded by the German Federal Cultural Foundation.

import sys
import signal
import argparse
import luna

##### parser

parser = argparse.ArgumentParser(description='''
Send TF-Luna LIDAR proximity distance measurements over OSC (default) or UDP
and optional "isThere" presence events to a ThingsBoard URL.

Distance format is cm integer or normalized float (inverted, 1 near to 0 far).

Message format: message [id] distance

  Ex. OSC: "/tfluna" distance

  Ex. UDP: "tfluna" distance

The default "tfluna" message can be overridden via --message and an additional
device identifier can be added with --id:

  $ --message /proximity --id 2
  OSC: "/proximity" 2 distance

Optionally, a boolean "isThere" event can be sent to a ThingsBoard URL whenever
someone moves in front or away from the sensor:

  $ --tb-url http://board.mydomain.com/api/v1/TOKEN/telemetry
  TB: {"isThere": 0}

The default "isThere" message name can be overridden via --tb-message:

  $ --tb-url http://board.mydomain.com/api/v1/TOKEN/telemetry \\
    --tb-message proximity
  TB: {"proximity": 0}

''', formatter_class=argparse.RawTextHelpFormatter)
parser.add_argument(
    "dev", type=str, nargs="?", metavar="DEV",
    default="/dev/ttyAMA0", help="serial port device, default: dev/ttyAMA0")
parser.add_argument(
    "-d", "--destination", dest="destination", metavar="HOST",
    default="127.0.0.1", help="destination hostname or IP address, default: 127.0.0.1")
parser.add_argument(
    "-p", "--port", type=int, dest="port", metavar="PORT",
    default="5005", help="destination port to send to, default: 5005")
parser.add_argument(
    "-i", "--interval", type=float, dest="interval", metavar="INTERVAL",
    default=0.1, help="read interval in seconds, default: 0.1")
parser.add_argument(
    "-e", "--epsilon", type=int, dest="epsilon", metavar="EPSILON",
    default=1, help="min distance change in cm to send a message, default: 2")
parser.add_argument(
    "-m", "--max-distance", type=int, dest="max_distance", metavar="MAX_DISTANCE",
    default=200, help="max allowed distance in cm, rest is clipped")
args = parser.add_argument(
    "-n", "--normalize", action="store_true", dest="normalize",
    help="send normalized values instead of cm: 1 near to 0 far (max distance)")
parser.add_argument(
    "-u", "--udp", action="store_true", dest="udp",
    default=False, help="send raw UDP message instead of OSC")
args = parser.add_argument(
    "--message", type=str, nargs=1, dest="message", metavar="MESSAGE",
    default=None, help="set OSC message address or UDP message text")
args = parser.add_argument(
    "--id", type=int, dest="devid", metavar="DEVID",
    default=None, help="set device identifier to include in message")
args = parser.add_argument(
    "--tb-url", type=str, dest="tb_url", metavar="TB_URL",
    default=None, help="send \"isThere\" bool message to a ThingsBoard url")
args = parser.add_argument(
    "--tb-message", type=str, nargs=1, dest="tb_message", metavar="TB_MESSAGE",
    default="isThere", help="set ThingsBoard \"isThere\" message name")
args = parser.add_argument(
    "-v", "--verbose", action="store_true", dest="verbose",
    help="enable verbose printing")

##### signal

# signal handler for nice exit
def sigint_handler(signum, frame):
    tfluna.stop()

##### main

if __name__ == '__main__':

    # parse
    args = parser.parse_args()

    # sensor
    try:
        tfluna = luna.TFLuna(dev=args.dev, verbose=args.verbose)
    except Exception as e:
        print(e)
        exit(1)
    tfluna.max_distance = args.max_distance
    tfluna.epsilon = args.epsilon
    tfluna.interval = args.interval
    tfluna.normalize = args.normalize
    tfluna.devid = args.devid
    if args.verbose:
        tfluna.print()

    # sender(s)
    if args.udp:
        sender = luna.UDPSender(addr=(args.destination, args.port))
        sender.message = " ".join(args.message) if args.message else "tfluna" 
        tfluna.add_sender(sender)
    else:
        sender = luna.OSCSender(addr=(args.destination, args.port))
        sender.address = " ".join(args.message) if args.message else "/tfluna"
        tfluna.add_sender(sender)
    if args.tb_url:
        sender = luna.TBSender(url=args.tb_url, threading=True)
        sender.message = "".join(args.tb_message)
        sender.verbose = args.verbose
        tfluna.add_sender(sender)
    if args.verbose:
        for sender in tfluna.senders:
            sender.print()

    # start
    signal.signal(signal.SIGINT, sigint_handler)
    try:
        tfluna.open()
        tfluna.start()
    finally:
        tfluna.close()
